### PR TITLE
bootloader/assets: support injecting bootloader assets in testing builds of snapd

### PR DESCRIPTION
**Pull request by bboozzoo**
Add support for injecting bootloader assets and corresponding snippets in test
builds of snapd. The code is only picked up when building with
`withbootassetstesting `tag.

The snippets are injected only when a specific marker file is present in the
same directory as the snapd binary. This way it is possible to control when a
snapd snap update will trigger an update of boot config assets of the grub
bootloader.

### CODE REVIEW

File: bootloader/withbootassettesting.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		panic(fmt.Sprintf("cannot inject boot config for asset: %v", err))
+	}
+	// bump he injected edition number
+	injectedEdition := internalEdition + 1
+
+	fmt.Printf("injecting grub boot assets for testing, edition: %v\n", injectedEdition)
+
+	lastSnippet := string(snippets[len(snippets)-1].Snippet)
+	injectedSnippet := lastSnippet + " bootassetstesting"
```
Comment (pedronis) [inline bootloader/withbootassettesting.go:86]: maybe what we inject should come from the marker file?

File: bootloader/assets/assetstesting.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+package assets
+
+import (
+	"fmt"
+)
+
+// InjectInternal injects an internal asset under the given name.
+func InjectInternal(name string, data []byte) {
+	fmt.Printf("injecting bootloader asset %q\n", name)
```
Comment (pedronis) [inline bootloader/assets/assetstesting.go:29]: shouldn't all the print use logger Debugf ?

File: bootloader/assets/assetstesting.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	registeredAssets[name] = data
+}
+
+func InternalSnippets(name string) []ForEditions {
```
Comment (pedronis) [inline bootloader/assets/assetstesting.go:33]: for consistency this should be called SnippetsForEditions ?

File: bootloader/assets/assetstesting.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	return registeredEditionSnippets[name]
+}
+
+// InjectSnippetForEditions injects a set of snippets under a given key.
+func InjectSnippetForEditions(name string, snippets []ForEditions) {
```
Comment (pedronis) [inline bootloader/assets/assetstesting.go:38]: s/Snippet/Snippets/

**Review by pedronis — COMMENTED** [review]
this is a bit hard to follow but looks reasonable, some comments

File: bootloader/withbootassettesting.go (hunk 1 above)
Comment (anonymouse64) [reply bootloader/withbootassettesting.go:86]: I agree, I think this is most useful if it could be used from a spread test to inject arbitrary things, not sure why we would need anything specific, but seems more useful than adding a kernel command line which isn't understood by anything :shrug:

File: bootloader/withbootassettesting_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+var _ = Suite(&withbootasetstestingTestSuite{})
+
+func (s *withbootasetstestingTestSuite) TestInjects(c *C) {
+	d := c.MkDir()
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "bootassetstesting"), nil, 0644), IsNil)
+	restore := bootloader.MockMaybeInjectOsReadlink(func(_ string) (string, error) {
+		return filepath.Join(d, "foo"), nil
+	})
+	defer restore()
+	restore = snapdenv.MockTesting(true)
```
Comment (anonymouse64) [inline bootloader/withbootassettesting_test.go:48]: is this really necessary? shouldn't snapdenv.IsTest return true anyways because we are in a test ?

**Review by anonymouse64 — COMMENTED** [review]
lgtm, but a few suggestions and I agree with the suggestions from @pedronis

**Review by anonymouse64 — APPROVED** [review]
(previous review was meant to be a +1 assuming the suggestions are implemented)

File: bootloader/assets/assetstesting.go (hunk 2 above)
Comment (bboozzoo) [reply bootloader/assets/assetstesting.go:29]: Added logger.Noticef() so that the logs would appear regardless of SNAPD_DEBUG

File: bootloader/withbootassettesting.go (hunk 1 above)
Comment (bboozzoo) [reply bootloader/withbootassettesting.go]: Done. The content becomes part of the kernel command line.

**Review by pedronis — APPROVED** [review]
thanks, the copyright year change is not applied
